### PR TITLE
fix tooltip message for 'var's. (#1001007)

### PR DESCRIPTION
**Pull request by andrey-ilinykh**

### CODE REVIEW

**Comment by sschaef** [comment]
ok to test

**Comment by sschaef** [comment]
The build failed because you introduced trailing whitespace: https://jenkins.scala-ide.org:8496/jenkins/job/ghprb-scala-ide-multi-conf/eclipse_platform=eclipse-luna/316/console

Also, please use `if (...)` format-style for the ifs.

File: org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala [hunk 1]
```diff
@@ -116,7 +116,21 @@ class ScalaHoverImpl extends ITextHover with ITextHoverExtension with ITextHover
...
+              Option("var " + sym.localName + ":" +gsym.tpe.resultType)
+            else
+              compiler.headerForSymbol(sym, tpe)
+          }
+          else  
+            compiler.headerForSymbol(sym, tpe)
+          
+            
+          
+        }).flatten
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: This produces a race condition. Symbols need to be accessed on the compiler thread. Wrap the yield block in:
```scala
compiler.asyncExec(...).getOrElse(None)()
```

File: org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala [hunk 2]
```diff
@@ -116,7 +116,21 @@ class ScalaHoverImpl extends ITextHover with ITextHoverExtension with ITextHover
 
         val scalaRegion = new Region(icu.sourceMap(viewer.getDocument.get.toCharArray).scalaPos(region.getOffset), region.getLength)
 
-        def typeInfo(t: Tree): Option[String] = (for (sym <- Option(t.symbol); tpe <- Option(t.tpe)) yield compiler.headerForSymbol(sym, tpe)).flatten
+        def typeInfo(t: Tree): Option[String] = (for (sym <- Option(t.symbol); tpe <- Option(t.tpe)) yield {
+          if( sym.isMethod && sym.isSourceMethod ){//this is info message for 'var's. We check if owner has both setter and getter
+            val gsym = sym.getterIn(sym.owner)
+            val ssym = sym.setterIn(sym.owner)
+            if( gsym != NoSymbol && ssym != NoSymbol)
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: This nested if can be removed by converting `gsym` and `ssym` to a def and then call them in the first if.

File: org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala [hunk 3]
```diff
@@ -116,7 +116,21 @@ class ScalaHoverImpl extends ITextHover with ITextHoverExtension with ITextHover
 
         val scalaRegion = new Region(icu.sourceMap(viewer.getDocument.get.toCharArray).scalaPos(region.getOffset), region.getLength)
 
-        def typeInfo(t: Tree): Option[String] = (for (sym <- Option(t.symbol); tpe <- Option(t.tpe)) yield compiler.headerForSymbol(sym, tpe)).flatten
+        def typeInfo(t: Tree): Option[String] = (for (sym <- Option(t.symbol); tpe <- Option(t.tpe)) yield {
+          if( sym.isMethod && sym.isSourceMethod ){//this is info message for 'var's. We check if owner has both setter and getter
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: This comment is unnecessary. Move the block to a def instead and give it a compelling name.

File: org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala [hunk 4]
```diff
@@ -116,7 +116,21 @@ class ScalaHoverImpl extends ITextHover with ITextHoverExtension with ITextHover
...
+            val gsym = sym.getterIn(sym.owner)
+            val ssym = sym.setterIn(sym.owner)
+            if( gsym != NoSymbol && ssym != NoSymbol)
+              Option("var " + sym.localName + ":" +gsym.tpe.resultType)
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: `localName` can return values with a space at the end (I forgot why). Please trim it.

Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: Replace `gsym.tpe.resultType` with `compiler.declPrinter(gsym.tpe.resultType)`.

**Comment by sschaef** [comment]
Actually, I think the bug should be fixed in `DeclarationPrinter.defString`. There is also a test: `DeclarationPrinterTest`. This would make the behavior change applicable everywhere and not just in the hover. Once it is fixed there, no further changes need to be done in the hover.

File: org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala (hunk 2 above)
Comment (andrey-ilinykh) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: But I need gsym later. Simply, I don't want to calculate it  twice.

**Comment by andrey-ilinykh** [comment]
I agree, DeclarationPrinter is the best place to fix. Let me look into it.

File: org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala (hunk 4 above)
Comment (andrey-ilinykh) [inline org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/hover/ScalaHoverImpl.scala]: What is the difference?

**Pull request closed without merging**